src/search3_rpc.erl

```diff
@@ -0,0 +1,207 @@
+            Default = construct_default(Analyzer, Stopwords),
+            #{prefix => Prefix, default => Default, per_field => Fields};
+        AnalyzerSpec ->
+            couch_log:notice("AnalyzerSpec ~p ", [AnalyzerSpec]),
+            #{prefix => Prefix, default => AnalyzerSpec}
+    end.
+
+construct_analyzer_spec(Analyzer) ->
+    {_, Name} = lists:keyfind(<<"name">>, 1, Analyzer),
+    Stopwords = case lists:keyfind(<<"stopwords">>, 1, Analyzer) of
+        false ->
+            [];
+        {_, Words} ->
+            Words
+    end,
+    #{name => Name, stopwords => Stopwords}.
+
+construct_per_fields(Analyzer) ->
+    Fields1 = case lists:keyfind(<<"fields">>, 1, Analyzer) of
+        false ->
+            [];
+        {_, {Fields}} ->
+            Fields
+    end,
+    MakeMapFun = fun
+        ({Field, PerFieldAnalyzer}, Map) when is_binary(PerFieldAnalyzer) ->
+            maps:put(Field, #{name => PerFieldAnalyzer}, Map);
+        ({Field, {PerFieldAnalyzer}}, Map) ->
+            AnalyzerSpec = construct_analyzer_spec(PerFieldAnalyzer),
+            maps:put(Field, AnalyzerSpec, Map)
+    end,
+    lists:foldl(MakeMapFun, #{}, Fields1).
+
+construct_default(Analyzer, Stopwords) ->
+    Default = case lists:keyfind(<<"default">>, 1, Analyzer) of
+        false ->
+            <<"standard">>;
+        {<<"default">>, Def} ->
+            Def
+    end,
+    #{name => Default, stopwords => Stopwords}.
+
+make_fields_map(Fields) when is_list(Fields) ->
+    FieldsMapFun = fun
+        ({Name, Value, {Options}}) ->
+            M1 = #{name => Name, value => fields_value(Value)},
+            Fields1 = options_list(Options),
+            M2 = maps:from_list(Fields1),
+            maps:merge(M1, M2)
+    end,
+    lists:map(FieldsMapFun, Fields).
+
+fields_value(Value) when is_binary(Value) ->
+    #{value => {string, Value}};
+fields_value(Value) when is_number(Value) ->
+    #{value => {double, Value}};
+fields_value(Value) when is_boolean(Value) ->
+    #{value => {bool, Value}}.
+
+% This function is required we need to convert <<"stored">>, <<"facet">>,
+% <<"analzyed">> into atoms for grpc.
+options_list(Options) when is_list(Options) ->
+    [{list_to_existing_atom(?b2l(Opt)), Val} || {Opt, Val} <- Options].
+
+construct_search_msg(Prefix, #index_query_args{}=QueryArgs) ->
+    #index_query_args{
+        q = Query,
+        limit = Limit,
+        bookmark = Bookmark,
+        stale = Stale,
+        sort = Sort,
+        partition = Partition,
+        counts = Counts,
+        ranges = Ranges,
+        drilldown = DrillDown,
+        include_fields = IncludeFields
+    } = QueryArgs,
+    Query1 = binary_to_list(Query),
+    SortArg = construct_sort_msg(Sort),
+    Msg = #{
+        index => Prefix,
+        query => Query1,
+        limit => Limit,
+        stale => Stale,
+        sort => SortArg
+    },
+    case construct_bookmark_msg(Bookmark) of
+        #{} -> ok;
+        Bookmark1 -> maps:put(bookmark, Bookmark1, Msg)
```

Comment:
It's a bit odd that we return an updated map or an ok for a bookmark. If the rest of the code is structured around this already, that's fine. Just wanted to pointed it out in case it wasn't intentional.

For updating, could do something like:

   ```#{} = Bookmark1 -> Msg#{bookmark := Bookmark1}```

Then also do we want to return the new updated Msg. Don't think we do it as the return from case is ignored.